### PR TITLE
Split libtorch.so back into libtorch_{cpu,cuda,hip}

### LINKED ISSUE #27215: LINK : fatal error LNK1102: out of memory (Windows CUDA CI) LNK
I've noticed this quite a few times, so I finally decided to file an issue about it.

A lot of py2-cuda9.0-cudnn7-windows-build are failing with 

```
23:54:12 caffe2\CMakeFiles\torch.dir\__\torch\csrc\cuda\comm.cpp.obj "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\lib\x64\cudart_static.lib" lib\cpuinfo.lib lib\foxi_loader.lib lib\cpuinfo.lib lib\c10_cuda.lib "C:\Program Files\NVIDIA Corporation\NvToolsExt\lib\x64\nvToolsExt64_1.lib" "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\lib\x64\cudart_static.lib" "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\lib\x64\cudart_static.lib" "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\lib\x64\cusparse.lib" "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\lib\x64\curand.lib" "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\lib\x64\curand.lib" "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\lib\x64\cudnn.lib" -WHOLEARCHIVE:C:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/build/lib/caffe2_protos.lib lib\clog.lib -WHOLEARCHIVE:C:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/build/lib/onnx.lib lib\onnx_proto.lib lib\libprotobuf.lib -WHOLEARCHIVE:C:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/build/lib/Caffe2_perfkernels_avx.lib -WHOLEARCHIVE:C:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/build/lib/Caffe2_perfkernels_avx2.lib -WHOLEARCHIVE:C:/Jenkins/workspace/caffe2-builds/py2-cuda9.0-cudnn7-windows-build/build/lib/Caffe2_perfkernels_avx512.lib lib\c10.lib "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\lib\x64\cudart_static.lib" "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\lib\x64\cufft.lib" "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\lib\x64\cublas.lib" kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib 
23:54:12    Creating library lib\torch.lib and object lib\torch.exp
23:54:12 LINK : fatal error LNK1102: out of memory
23:54:12 ninja: build stopped: subcommand failed.
```

https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-cuda9.0-cudnn7-windows-build/70519/console

I haven't seen this happen to PyTorch builds for some reason though

cc @ezyang @gchanan @zou3519 @jerryzh168 @peterjc123

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29731 Split libtorch.so back into libtorch_{cpu,cuda,hip}**

The new structure is that libtorch_cpu contains the bulk of our
code, and libtorch depends on libtorch_cpu and libtorch_cuda.

Some subtleties about the patch:
- There were a few functions that crossed CPU-CUDA boundary without API macros. I just added them, easy enough. An inverse situation was aten/src/THC/THCTensorRandom.cu where we weren't supposed to put API macros directly in a cpp file.
- DispatchStub wasn't getting all of its symbols related to static members on DispatchStub exported properly. I tried a few fixes but in the end gave up and rewrote the AVX/AVX2 registration mechanism to work the same way as the CUDA/HIP registration so I didn't have to deal with static members.  There were some shenanigans related to macro expansion. Also, there were some uses of the dispatch API which didn't include headers / declare definitions when they should have / declared definitions wrongly in anonymous namespaces, those have been fixed.
- torch/csrc/cuda/nccl.cpp was added to the wrong list of SRCS, now fixed (this didn't matter before because previously they were all in the same library)
- The dummy file for libtorch was brought back from the dead; it was previously deleted in #20774
- In an initial version of the patch, I forgot to make torch_cuda explicitly depend on torch_cpu. This lead to some very odd errors, most notably "bin/blob_test: hidden symbol `_ZNK6google8protobuf5Arena17OnArenaAllocationEPKSt9type_infom' in lib/l
ibprotobuf.a(arena.cc.o) is referenced by DSO" 
- A number of places in Android/iOS builds have to add torch_cuda explicitly as a library, as they do not have transitive dependency calculation working correctly

Fixes #27215 (as our libraries are smaller), and executes on
part of the plan in #29235.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18632773](https://our.internmc.facebook.com/intern/diff/D18632773)

### CODE REVIEW

**Comment by kostmo** [comment]
## CircleCI build failures summary
As of commit 699811b:
* **2/2** failures introduced in this PR
* **0/2** recognized as flaky



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=699811b0c84e427ab31463e2d891a2069ab58679):

| Job | Step | Log excerpt |
| --- | --- | --- |


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

**Comment by ezyang** [comment]
@xta0 I've cc'ed you as a reviewer because I edited some of the iOS test scripts to handle the library change

**Comment by xta0** [comment]
add @ljk53 . For iOS, there are two more files that need to be fixed:

1. `pytorch/ios/LibTorch.podspec`
2. `.circleci/scripts/binary_ios_upload.sh`

File: scripts/xcode_build.rb [hunk 1]
```diff
@@ -51,7 +51,7 @@
 
 # link static libraries
 target.frameworks_build_phases.clear
-libs = ['libc10.a', 'libclog.a', 'libnnpack.a', 'libeigen_blas.a', 'libcpuinfo.a', 'libpytorch_qnnpack.a', 'libtorch.a']
+libs = ['libc10.a', 'libclog.a', 'libnnpack.a', 'libeigen_blas.a', 'libcpuinfo.a', 'libpytorch_qnnpack.a', 'libtorch_cpu.a', 'libtorch.a']
```
Comment (xta0) [inline scripts/xcode_build.rb]: Do we still want to keep `libtorch.a`? If I understand correctly, we only need one which is `libtorch_cpu.a`

**Review by xta0 — CHANGES_REQUESTED** [review]
@ezyang this looks cool.  For iOS, there are two more files that are related, could you fix them as well

1. [LibTorch.podspec](https://github.com/pytorch/pytorch/blob/master/ios/LibTorch.podspec#L27)
2. [binary_ios_upload.sh](https://github.com/pytorch/pytorch/blob/master/.circleci/scripts/binary_ios_upload.sh#L17)

The first one is for the cocoapods build. The second one is our nightly job.

@IvanKobzarev @ljk53 Please take a look at Android.

File: scripts/xcode_build.rb (hunk 1 above)
Comment (ezyang) [reply scripts/xcode_build.rb]: libtorch.a is a stub library that doesn't contain any actual code. So it's harmless if we include it or not.

**Comment by ezyang** [comment]
@xta0 I fixed binary_ios_upload but I'm not too sure about the podspec. That one doesn't mention c10, which suggests to me that I don't need transitive dependencies here.

**Comment by xta0** [comment]
> @xta0 I fixed binary_ios_upload but I'm not too sure about the podspec. That one doesn't mention c10, which suggests to me that I don't need transitive dependencies here.

Oh I see. the one in podspec is telling XCode to apply whole archive to the `libtorch.a`. if we're going to use `libtorch_cpu.a`, then this should change as well.

**Comment by ezyang** [comment]
Added @ngimel for the semi-rewrite of DispatchStub.

**Comment by ljk53** [comment]
> @ezyang this looks cool. For iOS, there are two more files that are related, could you fix them as well
> 
> 1. [LibTorch.podspec](https://github.com/pytorch/pytorch/blob/master/ios/LibTorch.podspec#L27)
> 2. [binary_ios_upload.sh](https://github.com/pytorch/pytorch/blob/master/.circleci/scripts/binary_ios_upload.sh#L17)
> 
> The first one is for the cocoapods build. The second one is our nightly job.
> 
> @IvanKobzarev @ljk53 Please take a look at Android.

The android part needs to update is: https://github.com/pytorch/pytorch/blob/master/android/pytorch_android/CMakeLists.txt#L71

Without fixing it this PR probably won't pass android gradle test anyway...

@xta0 @IvanKobzarev we probably should use TorchConfig.cmake in iOS/Android projects after my fix for static libraries is landed: #29837 - but I'm not sure if it's something can be done before this PR lands.

File: scripts/xcode_build.rb (hunk 1 above)
Comment (xta0) [reply scripts/xcode_build.rb]: don't know what `libtorch.a` exactly contains. But if it doesn't contain any actual code, can we just skip it? Would there be any error without linking it?

Comment (ezyang) [reply scripts/xcode_build.rb]: I think we can skip it. I'm just kind of worried about what might happen if we do add code to the library at some point in the future: then we better not skip it!

Comment (xta0) [reply scripts/xcode_build.rb]: lol.  I guess you're right

**Review by xta0 — APPROVED** [review]
Looks good for iOS, needs to double check after this PR lands

**Comment by ezyang** [comment]
Thanks! I'll un-ack your review so other people look. Feel free to remove yourself from the reviewer list.

File: caffe2/CMakeLists.txt [hunk 2]
```diff
@@ -974,14 +998,20 @@ if (MSVC AND NOT BUILD_SHARED_LIBS)
...
+  target_compile_options(torch_cpu PRIVATE "-DONNX_BUILD_MAIN_LIB")
 endif()
 
+install(TARGETS torch_cpu EXPORT Caffe2Targets DESTINATION "${TORCH_INSTALL_LIB_DIR}")
+if (USE_CUDA)
+  install(TARGETS torch_cuda EXPORT Caffe2Targets DESTINATION "${TORCH_INSTALL_LIB_DIR}")
+elseif (USE_ROCM)
+  install(TARGETS torch_hip EXPORT Caffe2Targets DESTINATION "${TORCH_INSTALL_LIB_DIR}")
+endif()
 install(TARGETS torch EXPORT Caffe2Targets DESTINATION "${TORCH_INSTALL_LIB_DIR}")
```
Comment (ljk53) [inline caffe2/CMakeLists.txt]: @ezyang seems the "caffe2_interface_library(torch torch_library)" below won't add torch_cpu to the "--whole-archive" "--no-whole-archive" section thus causing some link error when building binaries with static linking.

I'm not sure if we can fix it inside "caffe2_interface_library" by adding "--whole-archive" around:
```
    target_link_libraries(${DST} INTERFACE
        $<TARGET_PROPERTY:${SRC},LINK_LIBRARIES>)
```

Or explicitly creating an interface library for torch_cpu?

Comment (ezyang) [reply caffe2/CMakeLists.txt]: Probably creating an interface library for torch_cpu is the right thing

**Pull request closed without merging**